File: 2023/java/D01P1.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class D01P1 {
    private static final int ZERO = 48;
    private static final int NINE = 57;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 1 - Part 1");

        BufferedReader reader = new BufferedReader(new FileReader(new File("../_input/day1.txt")));
        
        int total = 0;

        String line;
        while((line = reader.readLine()) != null) {
            total += get(line.trim());
        }

        System.out.println("Total: " + total);
    }

    private static int get(String str) {
        int first = -1;
        int last = -1;

        for(int c : str.chars().toArray()) {
            if(c >= ZERO && c <= NINE) {
                if(first == -1) first = c - ZERO;
                else {
                    last = c - ZERO;
                }
            }
        }

        if(last == -1) last = first;
        return (first * 10) + last;
    }
}

File: 2023/zig/d01p1.zig
const std = @import("std");

const ZERO: u8 = 48;
const NINE: u8 = 57;

pub fn main() anyerror!void {
    std.debug.print("Day 1 - Part 1\n", .{});

    const file = try std.fs.cwd().openFile("../_input/day1.txt", .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    const in = buf_reader.reader();

    var total: u32 = 0;

    var buffer: [1024]u8 = undefined;
    while (try in.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
        total += get(line);
    }

    std.debug.print("Total: {}\n", .{total});
}

fn get(str: []u8) u32 {
    var first: ?u32 = null;
    var last: ?u32 = null;

    for(str) |c| {
        if(c >= ZERO and c <= NINE) {
            if(first == null) {
                first = c - ZERO;
            }
            else {
                last = c - ZERO;
            }
        }
    }

    if(last == null) {
        last = first;
    }

    return (first.? * 10) + last.?;
}
